lib/kernels/src/cuda/transpose_kernels.cu

```diff
@@ -30,86 +33,99 @@ struct TransposeStrides {
 namespace Kernels {
 namespace Transpose {
 
+TransposePerDeviceState init_kernel(int num_dim, std::vector<int> const &perm) {
+
+  TransposePerDeviceState state;
+  state.num_dim = num_dim;
+  int const length = perm.size();
+
+  for (int i = 0; i < std::min(length, MAX_TENSOR_DIM); ++i) {
+    state.perm[i] = perm[i];
+  }
+
+  return state;
+}
+
+__global__ void transpose_simple_kernel(std::size_t volume,
+                                        float const *in_ptr,
+                                        float *out_ptr,
+                                        const TransposeStrides info,
+                                        float const beta) {
+  CUDA_KERNEL_LOOP(o_idx, volume) {
+    std::size i_idx = 0;
+    std::size t = o_idx;
+    for (int i = info.num_dim - 1; i >= 0; i--) {
+      coord_t ratio = t / info.out_strides[i];
+      t -= ratio * info.out_strides[i];
+      i_idx += ratio * info.in_strides[info.perm[i]];
+    }
+    out_ptr[o_idx] += out_ptr[o_idx] * beta + in_ptr[i_idx];
+  }
+}
+
 void forward_kernel(cudaStream_t stream,
-                    TransposePerDeviceState const *m,
+                    TransposePerDeviceState const &m,
                     float const *input_ptr,
                     float *output_ptr,
```

Comment:
If you're passing in `GenericTensorAccessorR/W` then you don't need these pointers